### ISSUE TITLE
1.19.4-rc1 crashes when searching in creative inventory for shields on worlds created on 1.19.3

### ISSUE THREAD

**Issue by V0g1** [anegostudios/VintageStory-Issues#3598]
### Game Version

v1.19.4-rc1

### Platform

Windows

### Modded

Modded

### SP/MP

Singleplayer

### Description

Game crashes when scrolling shield sections on worlds that were created on 1.19.3 version of the game. Creating new world on 1.19.4-rc1 resolves issue

### How to reproduce

1. Create world on 1.19.3
2. Upgrade to 1.19.4-rc1
3. Open world created on 1.19.3
4. When in creative mode type shield and scroll or just scroll everything page untill you go to shield section

### Logs

```Log```

Running on 64 bit Windows 10.0.19045.0 with 32553 MB RAM
Game Version: v1.19.4-rc.1 (Unstable)
17.02.2024 11:08:26: Critical error occurred
Loaded Mods: 1log4beams@1.0.0, aculinaryartillery@1.0.16, metalrecoverydon@1.0.1, bettercrates@1.7.0, bettertraders@0.0.5, butcheringfix@1.0.0, buzzwords@1.6.0, cellardoor@1.4.0, chiseltools@1.10.2, landformoverhaul@0.0.2, creativetapestries@1.0.0, driedpeat@1.0.3, durablewrenches@0.0.1, dye_recipes_fix@1.0.0, extendedcrafts@1.1.1, extrachests@1.8.0, flourbags@1.0.2, gwb@1.0.1, guicompositesettingsex@1.1.0, HangingOilLamps@1.0.1, healingwhilesleepingfix@1.0.1, hideplayername@2.0.5, hofe@0.0.3, hungrywakeup@1.0.1, immersivecorpsedrop@1.0.2, mannequinstand@1.0.3, metin2bgm@0.0.2, millwright@1.1.4, notradermoneylimit@1.0.0, pieontheshelf@1.0.0, potontheshelf@1.0.0, primitivesurvival@3.4.6, qptech@1.14.3, rackablefirestarter@1.0.0, she1fish@1.2.0, shepherdmod@1.0.2, steelbit1499@0.0.1, ssppvv@1.0.5, temporal_gears_stack@1.0.0, translocatorengineeringredux@1.4.9, game@1.19.4-rc.1, weedbegone@1.0.0, wrenchtip_disable@1.0.0, zoombuttonreborn@1.6.0, animalreproduction@1.0.0, abcsreborn@0.1.8, alloycalculator@1.0.0, egocaribautomapmarkers@3.1.0, beehivekiln@1.5.1, betterfirepit@1.1.4, betterjonasdevices@1.1.0, betterruins@0.3.3, helblockpick@2.3.0, bodyheatbar@1.0.1, branchcutter@1.0.1, carryon@1.7.4, commonlib@2.3.4, composter@1.1.0, driftersounds@1.0.0, extrachutes@1.0.0, extrainfo@1.7.0, fancyplanters@1.2.0, farmlanddropssoil@1.4.0, fixhandbookclutter@1.0.5, flowerfarming@1.0.0, fromgoldencombs@1.4.25, glowingprojectiles@1.1.3, hudclock@3.4.0, meteoricexpansion@1.2.5, morepiles@1.5.0, offhanddebuffpatch@0.5.0, ovenfuel@1.0.0, recyclebags@1.0.1, recycleclothes@1.0.0, recyclestorage@1.0.1, rustboundmagic@1.5.7, scarecrow@1.4.3, stonebakeoven@1.1.3, storagecontroller@1.0.6, storageoptions@1.0.2, tablet@0.1.4, th3dungeon@0.2.1-rc.1, th3expansion@1.2.0, useplanksinpitkiln@1.1.1, vtpp@1.1.1, creative@1.19.4-rc.1, vsimgui@0.3.5, survival@1.19.4-rc.1, whatbagwasthatagain@1.0.1, workbenchexpansion@1.8.0, worldconfiggui@1.0.1, xlib119@0.7.4-dev.119, barrelagedwood@1.0.0, bullseye@2.5.3, configlib@0.4.5, em@2.5.0, helvehammerext@1.6.0, maltiezbows@1.0.1, playercorpse@1.9.0, plumpkins@1.0.1, slanted_display_cases@1.3.2, stainedbeams@1.0.0, stonepiles@1.1.2, stonequarry@3.2.1, xskills119@0.7.5-dev.1191, yeet@4.0.1, bricklayers@2.5.0, itemmagnet@1.0.3
System.ArgumentNullException: Value cannot be null. (Parameter 'Unable to find a texture for texture code 'back' in item game:shield-metal. Giving up. Sorry.')
   at Vintagestory.Client.NoObf.ShapeTesselator.TesselateShapeElement(Int32 indexForLogging, MeshData meshdata, ShapeElement element, TesselationMetaData meta) in VintagestoryLib\Client\Render\Tesselation\ShapeTesselator.cs:line 394
   at Vintagestory.Client.NoObf.ShapeTesselator.TesselateShapeElements(MeshData meshdata, ShapeElement[] elements, TesselationMetaData meta) in VintagestoryLib\Client\Render\Tesselation\ShapeTesselator.cs:line 306
   at Vintagestory.Client.NoObf.ShapeTesselator.TesselateShapeElements(MeshData meshdata, ShapeElement[] elements, TesselationMetaData meta) in VintagestoryLib\Client\Render\Tesselation\ShapeTesselator.cs:line 306
   at Vintagestory.Client.NoObf.ShapeTesselator.TesselateShapeElements(MeshData meshdata, ShapeElement[] elements, TesselationMetaData meta) in VintagestoryLib\Client\Render\Tesselation\ShapeTesselator.cs:line 306
   at Vintagestory.Client.NoObf.ShapeTesselator.TesselateShapeElements(MeshData meshdata, ShapeElement[] elements, TesselationMetaData meta) in VintagestoryLib\Client\Render\Tesselation\ShapeTesselator.cs:line 306
   at Vintagestory.Client.NoObf.ShapeTesselator.TesselateShape(Shape shapeBase, MeshData& modeldata, Vec3f wholeMeshRotation, Vec3f wholeMeshOffset, Single wholeMeshScale, TesselationMetaData meta) in VintagestoryLib\Client\Render\Tesselation\ShapeTesselator.cs:line 302
   at Vintagestory.Client.NoObf.ShapeTesselator.TesselateShape(String type, AssetLocation name, CompositeShape compositeShape, MeshData& modeldata, ITexPositionSource texSource, Int32 generalGlowLevel, Byte climateColorMapIndex, Byte seasonColorMapIndex, Nullable`1 quantityElements, String[] selectiveElements) in VintagestoryLib\Client\Render\Tesselation\ShapeTesselator.cs:line 51
   at Vintagestory.Client.NoObf.ShapeTesselator.TesselateItem(Item item, MeshData& modeldata, ITexPositionSource texSource) in VintagestoryLib\Client\Render\Tesselation\ShapeTesselator.cs:line 645
   at Vintagestory.GameContent.ItemShield.GenMesh(ItemStack itemstack, ITextureAtlasAPI targetAtlas) in VSSurvivalMod\Item\ItemShield.cs:line 259
   at Vintagestory.GameContent.ItemShield.OnBeforeRender(ICoreClientAPI capi, ItemStack itemstack, EnumItemRenderTarget target, ItemRenderInfo& renderinfo) in VSSurvivalMod\Item\ItemShield.cs:line 153
   at Vintagestory.Client.NoObf.InventoryItemRenderer.GetItemStackRenderInfo(ClientMain game, ItemSlot inSlot, EnumItemRenderTarget target, Single dt) in VintagestoryLib\Client\Render\InventoryItemRenderer.cs:line 315
   at Vintagestory.Client.NoObf.InventoryItemRenderer.RenderItemstackToGui(ItemSlot inSlot, Double posX, Double posY, Double posZ, Single size, Int32 color, Single dt, Boolean shading, Boolean origRotate, Boolean showStackSize) in VintagestoryLib\Client\Render\InventoryItemRenderer.cs:line 194
   at Vintagestory.API.Client.GuiElementItemSlotGridBase.RenderInteractiveElements(Single deltaTime) in VintagestoryApi\Client\UI\Elements\Impl\Interactive\Inventory\GuiElementItemSlotGridBase.cs:line 509
   at Vintagestory.API.Client.GuiComposer.Render(Single deltaTime) in VintagestoryApi\Client\UI\GuiComposer.cs:line 716
   at Vintagestory.API.Client.GuiDialog.OnRenderGUI(Single deltaTime) in VintagestoryApi\Client\UI\Dialog\GuiDialog.cs:line 385
   at Vintagestory.Client.NoObf.GuiManager.OnRenderFrameGUI(Single deltaTime) in VintagestoryLib\Client\Systems\Gui\GuiManager.cs:line 318
   at Vintagestory.Client.NoObf.ClientEventManager.TriggerRenderStage(EnumRenderStage stage, Single dt) in VintagestoryLib\Client\Util\ClientEventManager.cs:line 186
   at Vintagestory.Client.NoObf.ClientMain.TriggerRenderStage(EnumRenderStage stage, Single dt) in VintagestoryLib\Client\ClientMain.cs:line 807
   at Vintagestory.Client.NoObf.ClientMain.RenderToDefaultFramebuffer(Single dt) in VintagestoryLib\Client\ClientMain.cs:line 1011
   at Vintagestory.Client.ScreenManager.Render(Single dt) in VintagestoryLib\Client\ScreenManager.cs:line 676
   at Vintagestory.Client.ScreenManager.OnNewFrame(Single dt) in VintagestoryLib\Client\ScreenManager.cs:line 651
   at Vintagestory.Client.NoObf.ClientPlatformWindows.window_RenderFrame(FrameEventArgs e) in VintagestoryLib\Client\ClientPlatform\GameWindow.cs:line 88
   at OpenTK.Windowing.Desktop.GameWindow.Run()
   at Vintagestory.Client.ClientProgram.Start(ClientProgramArgs args, String[] rawArgs) in VintagestoryLib\Client\ClientProgram.cs:line 324
   at Vintagestory.Client.ClientProgram.<>c__DisplayClass9_0.<.ctor>b__1() in VintagestoryLib\Client\ClientProgram.cs:line 128
   at Vintagestory.ClientNative.CrashReporter.Start(ThreadStart start) in VintagestoryLib\Client\ClientPlatform\ClientNative\CrashReporter.cs:line 93

**Comment by Craluminum2413** [comment]
@V0g1 Does it happen without any mods installed?

**Comment by V0g1** [comment]
Just checked - no. Only happens with mods enabled

**Comment by V0g1** [comment]
Found issue - it's Primitive Survival mod. Will contact mod author :)

**Comment by SpearAndFang** [comment]
Much more likely to be the vanilla remapper tool that pops up post upgrade. It also seems to have broken one of my mods other assets.  Let's call this a vanilla issue.

**Comment by tyronx** [comment]
I'm not able to reproduce this though. Created a 1.19.3 world with primitive survival, then upgrade to 1.19.4-rc.2 and searched for shield in creative inv. No crash.

**Comment by radfast** [comment]
At face value, the reported issue is this:
System.ArgumentNullException: Value cannot be null. (Parameter 'Unable to find a texture for texture code 'back' in item game:shield-metal. Giving up. Sorry.')

`shield-metal` is a vanilla item, though I guess a mod might have modded it...
It has versions with different metals, metal type is stored in attributes.  For the Creative Inventory these different versions of the shield are **generated at runtime** not stored in the save game.

The different metals (in vanilla) should be: `"tinbronze", "bismuthbronze", "blackbronze", "gold", "silver", "iron", "meteoriciron", "steel"`

The texture `back` - which is erroring - should correspond to the metal and should be the metal name, found in folder `"block/metal/plate/"`.  Obviously in vanilla that folder has textures for tinbronze, bizmuthbronze, blackbronze, gold, silver, iron, meteoriciron and steel.

The crash log shows the modded game is either missing one of those vanilla textures, or a mod has added a new metal shield version without the matching plate texture.

I don't see any way this could occur in unmodded vanilla.   Also doesn't look to have any connection with Primitive Survival mod, that looks a wrong result @V0g1.  Do you have any mod in that long list which: adds new metals or removes a metal?
Can we have a screenshot of the Creative Inventory, shields section, if you are able to get it to display without a crash?

Of course it's possible something else different is going on here, no connection with shields just causing this as a side-effect.  But the crash log is very specific.

**Comment by radfast** [comment]
See also [similar issue reported on Discord](https://discord.com/channels/302152934249070593/1212309109924298784), some overlapping mods in the list

**Comment by radfast** [comment]
On further careful examination and testing - tested with a full set of the mods which caused the similar issue on Discord - I'm coming to the provisional conclusion that _no specific mod caused this_, it's simply the overall amount of modded content meant that shields had no space for the various textures they need.  (Shields are kind of last to the party, when it comes to inclusion in the texture atlas, because their meshes are created at runtime only when the Creative Inventory or the Handbook are first opened - almost every other block and item creates its meshes during game startup)

Specific triggers for the issue may include textureAtlas size settings in ClientSettings, and which other blocks are around in the place where the player first joins the game  (a small handful of other block types also upload textures at runtime, like the shields do).  It's hard to be sure without having the exact world, position in that world, and settings, of a person for whom the issue is happening.

Anyhow, in 1.19.5 we can mitigate this: we can log an error and draw the shield with a missing texture on the affected part, instead of crashing.

**Comment by derg4** [comment]
I have the same issue with my modset in 1.19.3 stable, I was messing around trying to figure out what mod was causing the issue and found a collection of 12 disparate mods, any of which when removed would fix the problem. My go-to way to reproduce the issue is to open the handbook, search "shield" and scroll down. This almost instantly causes a crash. I'm sure it's the same as using the creative inventory.

[vs_shieldcrash.txt](https://github.com/anegostudios/VintageStory-Issues/files/14467066/vs_shieldcrash.txt)

I just want to add that when I launch Vintage Story I use a shortcut which passes a --dataPath argument, but I found that if I launch the game without the argument (but with the same mods, just copied to the default VintageStoryData folder) I do not crash. (For good measure I did this with a fresh install and on a fresh map too). I have no idea why, but I hope it helps.

I also want to say that I'm finding this issue playing on a multiplayer server as well, which made me want to look into it. I'm one of the few players that's having this issue, and a friend who logs in in the same area never has had this crash.

> Specific triggers for the issue may include textureAtlas size settings in ClientSettings...

I was able to fix my consistent crashing in a separate way also, which was increasing the textureAtlas size in clientsettings.json based off of this comment. Thanks!
```
    "maxTextureAtlasWidth": 4096,
    "maxTextureAtlasHeight": 4096,
```
